jetty-security/src/main/java/org/eclipse/jetty/security/DynamicAuthenticator.java

```diff
@@ -0,0 +1,282 @@
+
+import org.eclipse.jetty.http.pathmap.MappedResource;
+import org.eclipse.jetty.http.pathmap.PathMappings;
+import org.eclipse.jetty.http.pathmap.ServletPathSpec;
+import org.eclipse.jetty.server.Authentication;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.URIUtil;
+import org.eclipse.jetty.util.component.ContainerLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+import org.eclipse.jetty.util.security.Constraint;
+
+/**
+ * Used as a mapping of {@link org.eclipse.jetty.http.pathmap.PathSpec} to {@link Authenticator}.
+ *
+ * This allows you to register different authentication mechanisms for different paths.
+ */
+public class DynamicAuthenticator extends ContainerLifeCycle implements Authenticator
+{
+    private static final Logger LOG = Log.getLogger(DynamicAuthenticator.class);
+
+    private final PathMappings<AuthenticatorEntry> _mappings = new PathMappings<>();
+
+    public void addMapping(String pathSpec, Authenticator authenticator)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator));
+    }
+
+    public void addMapping(String pathSpec, Authenticator authenticator, LoginService loginService)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator, loginService));
+    }
+
+    public void addMapping(String pathSpec, Authenticator authenticator, LoginService loginService, IdentityService identityService)
+    {
+        addMapping(pathSpec, new AuthenticatorEntry(authenticator, loginService, identityService));
+    }
+
+    private void addMapping(String pathSpec, AuthenticatorEntry entry)
+    {
+        _mappings.put(new ServletPathSpec(pathSpec), entry);
+        addBean(entry);
+    }
+
+    @Override
+    public void setConfiguration(AuthConfiguration configuration)
+    {
+        for (MappedResource<AuthenticatorEntry> mapping : _mappings)
+        {
+            AuthenticatorEntry entry = mapping.getResource();
+            DynamicConfiguration config = new DynamicConfiguration(configuration, entry);
+            entry.getAuthenticator().setConfiguration(config);
+        }
+    }
+
+    @Override
+    public String getAuthMethod()
+    {
+        return Constraint.__DYNAMIC_AUTH;
+    }
+
+    @Override
+    public void prepareRequest(ServletRequest request)
+    {
+        Authenticator authenticator = getAuthenticator(request);
+        if (authenticator == null)
+            return;
+
+        authenticator.prepareRequest(request);
+    }
+
+    @Override
+    public Authentication validateRequest(ServletRequest request, ServletResponse response, boolean mandatory) throws ServerAuthException
+    {
+        final Request baseRequest = Objects.requireNonNull(Request.getBaseRequest(request));
+        final Response baseResponse = baseRequest.getResponse();
+
+        Authenticator authenticator = getAuthenticator(request);
+        if (authenticator == null)
```

Comment:
If there isn't an authenticator that matches the path of the request, maybe this is a path that doesn't require authentication?  How would I configure a webapp to have some paths that have security constraints and require authentication, and others that don't? Currently eg the FormAuthenticator uses DeferredAuthentication to handle this scenario.